modules/ssh/ssh.go

```diff
@@ -10,178 +9,179 @@ import (
-			defer func() {
-				if err = ch.Close(); err != nil {
-					log.Error("Close: %v", err)
-				}
-			}()
-			for req := range in {
-				payload := cleanCommand(string(req.Payload))
-				switch req.Type {
-				case "exec":
-					cmdName := strings.TrimLeft(payload, "'()")
-					log.Trace("SSH: Payload: %v", cmdName)
-
-					args := []string{"serv", "key-" + keyID, "--config=" + setting.CustomConf}
-					log.Trace("SSH: Arguments: %v", args)
-					cmd := exec.Command(setting.AppPath, args...)
-					cmd.Env = append(
-						os.Environ(),
-						"SSH_ORIGINAL_COMMAND="+cmdName,
-						"SKIP_MINWINSVC=1",
-					)
-
-					stdout, err := cmd.StdoutPipe()
-					if err != nil {
-						log.Error("SSH: StdoutPipe: %v", err)
-						return
-					}
-					stderr, err := cmd.StderrPipe()
-					if err != nil {
-						log.Error("SSH: StderrPipe: %v", err)
-						return
-					}
-					input, err := cmd.StdinPipe()
-					if err != nil {
-						log.Error("SSH: StdinPipe: %v", err)
-						return
-					}
-
-					// FIXME: check timeout
-					if err = cmd.Start(); err != nil {
-						log.Error("SSH: Start: %v", err)
-						return
-					}
-
-					err = req.Reply(true, nil)
-					if err != nil {
-						log.Error("SSH: Reply: %v", err)
-					}
-					go func() {
-						_, err = io.Copy(input, ch)
-						if err != nil {
-							log.Error("SSH: Copy: %v", err)
-						}
-					}()
-					_, err = io.Copy(ch, stdout)
-					if err != nil {
-						log.Error("SSH: Copy: %v", err)
-					}
-					_, err = io.Copy(ch.Stderr(), stderr)
-					if err != nil {
-						log.Error("SSH: Copy: %v", err)
-					}
-
-					if err = cmd.Wait(); err != nil {
-						log.Error("SSH: Wait: %v", err)
-						return
-					}
-
-					_, err = ch.SendRequest("exit-status", false, []byte{0, 0, 0, 0})
-					if err != nil {
-						log.Error("SSH: SendRequest: %v", err)
-					}
-					return
-				default:
-				}
-			}
-		}(reqs)
+		return 1
 	}
+
+	return waitStatus.ExitStatus()
 }
 
-func listen(config *ssh.ServerConfig, host string, port int) {
-	listener, err := net.Listen("tcp", host+":"+com.ToStr(port))
+func shellEscape(args []string) string {
+	var data = make([]string, len(args))
+	for k, v := range args {
+		data[k] = v
+
+		// This is a borderline dirty hack. It's designed to only escape
+		// strings which *might* need to be escaped. It uses a very
+		// limited character set so everything that needs to be quoted
+		// will be but we can ignore some simple cases to make it easier
+		// to parse in the ssh hook. We allow alpha-numeric characters
+		// along with ., _, and -
+		idx := strings.IndexFunc(v, func(r rune) bool {
+			return !unicode.IsLetter(r) && !unicode.IsDigit(r) && !strings.ContainsRune(".-_", r)
+		})
+		if idx > -1 || v == "" {
+			data[k] = "'" + strings.Replace(data[k], "'", "'\"'\"'", -1) + "'"
+		}
+	}
+	return strings.Join(data, " ")
+}
+
+func sessionHandler(session ssh.Session) {
+	keyID := session.Context().Value(giteaKeyID).(int64)
+
+	command := shellEscape(session.Command())
```

Comment:
This is unnecessary and potentially incorrect - cmd.Env will do this correctly for us.

Reply:
updated. Thanks :)

Reply:
Seems these changes broke tests. I will revert and attempt again.

Reply:
As I commented below, I believe you are mistaken about the use of shellEscape here. session.Command() returns a slice, not a single value. Because we need to fit that into a single environment variable, we need to escape some values so it looks like a proper shell command. This is perhaps related to https://github.com/gliderlabs/ssh/pull/93. Maybe it would make more sense to use the raw command (which isn't in the library yet, but I can add if it's needed).